src/services/formatting/rules.ts

```diff
@@ -725,17 +725,7 @@ namespace ts.formatting {
     }
 
     function isEndOfDecoratorContextOnSameLine(context: FormattingContext): boolean {
-        return context.TokensAreOnSameLine() &&
-            hasDecorators(context.contextNode) &&
-            nodeIsInDecoratorContext(context.currentTokenParent) &&
-            !nodeIsInDecoratorContext(context.nextTokenParent);
-    }
-
-    function nodeIsInDecoratorContext(node: Node): boolean {
-        while (isExpressionNode(node)) {

```

Comment:
`isExpressionNode` checks `.parent` in some cases to determine whether the node is, in fact, part of an expression. However, in the linked issue this is being used in tandem with `ChangeTracker`, which often means new synthetic nodes are added to the tree that don't have a valid `.parent`. Since TypeScript's own `Node` type lies about `parent (it's not always set despite not having an `| undefined` in its definition), functions that depend on `.parent` aren't always reliable except when working with original parse tree nodes.

One potential fix is to leave this in and use `while (node && isExpressionNode(node))` to guard against a missing parent. In addition, it might be better to use the more general `isExpression` test instead of `isExpressionNode`, since `isExpression` doesn't check `.parent` (at the cost of potentially misattributing `Identifier`, `this`, `true`, `false`, or `null` in a non-expression position as an expression). Given how this function is used, `isExpression` (along with the `node && ...` guard) is probably fine.

Reply:
> However, in the linked issue this is being used in tandem with ChangeTracker, which often means new synthetic nodes are added to the tree that don't have a valid .parent.

This is the root cause :). We can try to add more tests to [formattingDecorators](https://github.com/microsoft/TypeScript/blob/main/tests/cases/fourslash/formattingDecorators.ts), or just use `node && ...` to skip this rule for synthetic nodes. Both ways are ok for me :)

Reply:
@rbuckton I've changed the condition to `node && ...`.